### ISSUE TITLE
esbuild: Application rebuilding after files wrongfully were detected as changed

### ISSUE THREAD

**Issue by wartab** [angular/angular-cli#25197]
### Command

serve

### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### Description

Sometimes, files will be detected as changed (I assume by chokidar), even if they haven't actually changed. This causes the application to rebuild. 

This is not the case when not using esbuild. 

Those file changes can be caused by multiple things, like git or WebStorm randomly deciding to writing files to disk that haven't been changed. 

### Minimal Reproduction

run ng serve

One way of replicating this behaviour in WebStorm is to press Ctrl + S even if the file hasn't changed.

### Your Environment

```text
_                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/


Angular CLI: 16.0.1
Node: 16.18.0
Package Manager: npm 9.6.4
OS: win32 x64

Angular: 16.0.1
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, localize, material
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1600.1
@angular-devkit/build-angular   16.0.1
@angular-devkit/core            16.0.1
@angular-devkit/schematics      16.0.1
@schematics/angular             16.0.1
rxjs                            7.8.1
typescript                      5.0.4
```

**Comment by alan-agius4** [comment]
@wartab, this should not be any different from the default webpack based builder. When a file is saved, even if the contents did not change a rebuild is triggered.

**Comment by wartab** [comment]
I just tested it again. I can confirm what I reported. The behaviour is different with esbuild.

**Comment by wartab** [comment]
I'm not sure how to help out with this. It's pretty easy to replicate on my end, so if you can tell me how to proceed, let me know.

**Comment by clydin** [comment]
As mentioned, the default Webpack-based builder also will rebuild if a file is altered but does not have any content changes.

The following steps can be used to demonstrate this:
1) Execute `ng build --watch`. Output:
```
✔ Browser application bundle generation complete.
✔ Index html generation complete.

Initial Chunk Files           | Names         |  Raw Size | Estimated Transfer Size
main.bb3905875abf73da.js      | main          | 123.30 kB |                36.98 kB
polyfills.17229f07c3c9bed8.js | polyfills     |  32.99 kB |                10.67 kB
runtime.979e5903ddb9e524.js   | runtime       | 898 bytes |               509 bytes
styles.ef46db3751d8e999.css   | styles        |   0 bytes |                       -

                              | Initial Total | 157.17 kB |                48.14 kB

Build at: 2023-05-15T16:15:38.575Z - Hash: 4c850a06340d6159 - Time: 9746ms
```

2) In another console execute`touch src/main.ts`.  Output:
```
✔ Browser application bundle generation complete.
✔ Index html generation complete.

4 unchanged chunks

Build at: 2023-05-15T16:15:55.232Z - Hash: 4c850a06340d6159 - Time: 423ms
```

As shown above, a rebuild is performed even though no input content is actually changed.

**Comment by psarno** [comment]
I can confirm I am seeing this also, but it seems totally random 😕

I had VS Code opened and cleared the terminal so I could keep an eye on it.

All I did was click on my ".gitignore" file, just _click it_ not change it, and I got "Changes detected.  Rebuilding ...".

Then I cleared the terminal and tried clicking other files, couldn't repro it ... until I clicked angular.json.   Then it rebuilt again.

If I clear the terminal and try clicking these files again to repro, it won't rebuild.

No modifications were made to these files at all, it simply displayed them in the editor and this caused esbuild to trigger another rebuild.

Next, I went to clear the terminal again, so I clicked the three dots to get to the "Clear Terminal" window.

Just _clicking the three dots_ again caused it to rebuild.  Just the opening of the menu.

![image](https://github.com/angular/angular-cli/assets/18640860/bc9c41fe-bd74-429d-a086-59b6a832bc61)

**Comment by d9j** [comment]
it's quite annoying problem. sometimes it seems randomly rebuilding while viewing front end in the browser!

**Comment by leo6104** [comment]
Might be related to https://github.com/angular/angular-cli/commit/7155cbe5b2991f986c0683d16a6fc4d9b411e47b

**Comment by wartab** [comment]
> Might be related to [7155cbe](https://github.com/angular/angular-cli/commit/7155cbe5b2991f986c0683d16a6fc4d9b411e47b)

It does fix it, indeed. That said, and I realise it's a bit nitpicky, but that's not exactly the behaviour we had when not using esbuild, but I'm sure it doesn't affect the vast majority of projects:

- Directories leading with a dot shouldn't be ignored if they are inside the src/ directory, in my opinion.
- Regarding the nitpick: without esbuild, even outside the src/ directory, Angular tracks changes for files that are imported by TypeScript.

**Comment by d9j** [comment]
> Might be related to [7155cbe](https://github.com/angular/angular-cli/commit/7155cbe5b2991f986c0683d16a6fc4d9b411e47b)

The fix indeed seems correct. but why not allow user to customize behaviour  of esbuild .e.g have es build config inside project  ?

**Comment by clydin** [comment]
Addressed via #26182
